controller/rpc/scanner.go

```diff
@@ -15,90 +15,147 @@ import (
 )
 
 type scannerAct struct {
-	scanner  *share.CLUSScanner
-	scanning uint32
+	scanner            *share.CLUSScanner
+	activeScannerTasks int // Indicate how much scanner task is currently run
 }
 
-var scanners map[string]*scannerAct = make(map[string]*scannerAct)
-var scanMutex sync.RWMutex
+type ScannerManager struct {
+	activeScanners map[string]*scannerAct
+	maxConns       int
+	mutex          sync.RWMutex
+	availableCh    chan struct{}
+}
+
+var ScannerMgr *ScannerManager
+
+// scannerChannelCapacity ensures that the availableCh channel has sufficient capacity to handle task signals without blocking.
+var scannerChannelCapacity = 15
+var getAvailableScannerTimeout = time.Minute * 3
+
+func NewScannerManager(maxConns int) {
+	ScannerMgr = &ScannerManager{
+		activeScanners: make(map[string]*scannerAct),
+		maxConns:       maxConns,
+		availableCh:    make(chan struct{}, scannerChannelCapacity*maxConns),
+	}
+}
 
-func AddScanner(scanner *share.CLUSScanner) {
-	scanMutex.Lock()
-	defer scanMutex.Unlock()
-	scanners[scanner.ID] = &scannerAct{scanner: scanner}
+func (mgr *ScannerManager) signalAvailability() {
+	select {
+	case mgr.availableCh <- struct{}{}:
+		// Singal the pending task
+	default:
+		// Signal already pending, do not block
+	}
+}
+
+func (mgr *ScannerManager) AddScanner(scanner *share.CLUSScanner) {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	mgr.activeScanners[scanner.ID] = &scannerAct{scanner: scanner, activeScannerTasks: 0}
+	mgr.signalAvailability()
 }
 
-func RemoveScanner(scannerID string) {
-	scanMutex.Lock()
-	s, ok := scanners[scannerID]
+func (mgr *ScannerManager) RemoveScanner(scannerID string) {
+	mgr.mutex.Lock()
+	s, ok := mgr.activeScanners[scannerID]
 	if ok {
-		delete(scanners, scannerID)
+		delete(mgr.activeScanners, scannerID)
 	}
-	scanMutex.Unlock()
+	mgr.mutex.Unlock()
 
 	// endpoint is also the key
 	if s != nil {
-		endpoint := getScannerEndpoint(s.scanner)
+		endpoint := mgr.getScannerEndpoint(s.scanner)
 		cluster.DeleteGRPCClient(endpoint)
 	}
 }
```

Comment:
I can see the availability will increase when we have a new scanner, but do we need to decrease it when a scanner is removed?